### PR TITLE
Make sure we don't scan files that can not be accessed

### LINKED ISSUE #1965: can't delete directory
### Steps to reproduce
i have a script which transforms an project into static html. There was an failure in the script and it created a folder with an "\n" in the middle.

So the link would be something like:
https://nextcloud.td/index.php/apps/files?dir=/xxx%0Ayyy

The Nextcloud-Server 10.0.1 shows the path, but you can not rename or delete the folder.

Most problematic thing at all is that  the behaviour of the client (testet on Windows 2.2.4 breaks the sync completely.
 
### Expected behaviour
First: you should always rename or delete a folder or file.
Second: The sync should display the problem, but so sync all the rest.

### Actual behaviour
Can't delete or rename the folder in the frontend.

Can delete the folder in user/files/ but the server don't remove the entries.

Sync-Client breaks.
### Server configuration
**Operating system**:
Linux

**Web server:**
apache

**Database:**
mysql 5.5

**PHP version:**
php 5.5

**Nextcloud version:** (see Nextcloud admin page)
10.0.1

**Updated from an older Nextcloud/ownCloud or fresh install:**
fresh installation on a hosting server

```
Login as admin user into your Nextcloud and access 
http://example.com/index.php/settings/integrity/failed 
paste the results here.
```
No errors have been found.
</details>

**The content of config/config.php:**
<details>
<summary>Config report</summary>

```
{
    "system": {
        "instanceid": "ocwzn2u9mqmo",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "xyz.de"
        ],
        "datadirectory": "\/homepages\/htdocs\/nextcloud_data",
        "overwrite.cli.url": "https:\/\/xyz.de\/cloud",
        "dbtype": "mysql",
        "version": "9.1.1.5",
        "dbname": "db4",
        "dbhost": "db.com",
        "dbport": "",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "logtimezone": "Europe\/Berlin",
        "installed": true,
        "filesystem_check_changes": 1,
        "debug": false,
        "mail_from_address": "info",
        "mail_smtpmode": "php",
        "mail_domain": "xyz.de",
        "check_for_working_htaccess": true
    }
}


**Are you using external storage, if yes which one:** local/smb/sftp/...
no, only local attached storage

**Are you using encryption:** yes/no
no encryption

#### Nextcloud log (data/nextcloud.log)
<details>
<summary>Nextcloud log</summary>
```
Insert your Nextcloud log here
```
{"reqId":"WBcpsdmgP44AACt4uhsAAAAO","remoteAddr":"2a02:8071:278b:5200:2d5f:cf25:3193:d6","app":"webdav","message":"Exception: {\"Message\":\"HTTP\\\/1.1 400 Der Dateiname enth\\u00e4lt mindestens ein ung\\u00fcltiges Zeichen\",\"Exception\":\"OCA\\\\DAV\\\\Connector\\\\Sabre\\\\Exception\\\\InvalidPath\",\"Code\":0,\"Trace\":\"#0 \\\/homepages\\\/htdocs\\\/nextcloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(903): OCA\\\\DAV\\\\Connector\\\\Sabre\\\\ObjectTree->getNodeForPath('backups\\\/grundof...')\\n#1 \\\/homepages\\\/htdocs\\\/nextcloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/CorePlugin.php(334): Sabre\\\\DAV\\\\Server->getPropertiesForPath('backups\\\/grundof...', Array, 1)\\n#2 [internal function]: Sabre\\\\DAV\\\\CorePlugin->httpPropFind(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#3 \\\/homepages\\\/14\\\/d541580074\\\/htdocs\\\/nextcloud\\\/3rdparty\\\/sabre\\\/event\\\/lib\\\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\\n#4 \\\/homepages\\\/htdocs\\\/nextcloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(459): Sabre\\\\Event\\\\EventEmitter->emit('method:PROPFIND', Array)\\n#5 \\\/homepages\\\/htdocs\\\/nextcloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(248): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#6 \\\/homepages\\\/htdocs\\\/nextcloud\\\/apps\\\/dav\\\/appinfo\\\/v1\\\/webdav.php(60): Sabre\\\\DAV\\\\Server->exec()\\n#7 \\\/homepages\\\/htdocs\\\/nextcloud\\\/remote.php(165): require_once('\\\/homepages\\\/d...')\\n#8 {main}\",\"File\":\"\\\/homepages\\\/htdocs\\\/nextcloud\\\/apps\\\/dav\\\/lib\\\/Connector\\\/Sabre\\\/ObjectTree.php\",\"Line\":129,\"User\":\"xxx\"}","level":4,"time":"2016-10-31T12:23:30+01:00","method":"PROPFIND","url":"\/cloud\/remote.php\/webdav\/img\/projekte\/2014-11%0A2014","user":"xxx"}
</details>

### PR DESCRIPTION
**Pull request by nickvergessen**
### Steps
1. create a folder on the file system:
```
mkdir "data/admin/files/te
st"
```
2. Run `files:scan admin`
3. Try to delete the folder after realizing your problem

With this fix the file is not added on scanning.

@icewind1991 @MorrisJobke @LukasReschke 

Fix #1965

### CODE REVIEW

**Comment by MorrisJobke** [comment]
```
1) OCA\Files_Sharing\Tests\External\ScannerTest::testScan
Your test case is not allowed to access the database.
2) OCA\Files_Sharing\Tests\External\ScannerTest::testScanFile
Your test case is not allowed to access the database.
```

**Comment by nickvergessen** [comment]
I will fix that, once @icewind1991 confirmed this is the way to go.

**Review by icewind1991 — CHANGES_REQUESTED** [review]

File: lib/private/Files/Cache/Scanner.php [hunk 1]
```diff
@@ -132,6 +132,24 @@ protected function getData($path) {
 	 */
 	public function scanFile($file, $reuseExisting = 0, $parentId = -1, $cacheData = null, $lock = true) {
 
+		if (!\OC::$server->getDatabaseConnection()->supports4ByteText()) {
```
Comment (nickvergessen) [inline lib/private/Files/Cache/Scanner.php]: That was exactly what I wondered aswell, but I think the cache is inside the view, so this looks shitty from dependency pov. But should we just move it to a separate helper or something?

Comment (nickvergessen) [inline lib/private/Files/Cache/Scanner.php]: @icewind1991 should we simply move all the checks from the View into `IStorage::verifyPath` and in the view then simply resolve the path to the storage and call the method?

**Comment by nickvergessen** [comment]
@icewind1991 moved the validation from view to storage so it's in one place only

**Comment by MorrisJobke** [comment]
Tested and works 👍 

@rullzer @icewind1991 Please review :)

**Comment by rullzer** [comment]
Fine by me 👍 But lets have our fs guru have the last vote. Summoning @icewind1991

**Comment by icewind1991** [comment]
:+1:

**Review by icewind1991 — APPROVED** [review]

**Comment by MorrisJobke** [comment]
@nickvergessen Could you open the backport PR? Thanks

**Pull request merged**